<h1 align="center">💳 Credit Card Fraud</h1>

# 🎯 Project Goal

The goal is to identify **fraudulent credit card transactions in real-time**.

The dataset contains transaction details such as:
- ⏱️ **Time**: when the transaction occurred
- 💰 **Amount**: the transaction value
- 🧾 Other anonymized features (for example, V1 to V28 in the common dataset)


- **1 The Data**

In [1]:
#The dataset provides the raw information needed to train the machine learning model to distinguish between fraudulent and non-fraudulent transactions.
import kagglehub

# Download latest version
path = kagglehub.dataset_download("mlg-ulb/creditcardfraud")

print("Path to dataset files:", path)

Path to dataset files: C:\Users\KOVVO\.cache\kagglehub\datasets\mlg-ulb\creditcardfraud\versions\3


- **import some necessary libraries**

In [2]:
# import some necessary libraries
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import plotly.express as px

In [3]:
# read the data
data = pd.read_csv("C://Users//KOVVO//.cache//kagglehub//datasets//mlg-ulb//creditcardfraud//versions//3//creditcard.csv")

data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
# check the datatypes of the data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [5]:
# check the missing values
data.isnull().sum()*100

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [6]:
# check the statistics of the data 
data.describe().T

count          mean           std         min           25%  \
Time    284807.0  9.481386e+04  47488.145955    0.000000  54201.500000   
V1      284807.0  1.168375e-15      1.958696  -56.407510     -0.920373   
V2      284807.0  3.416908e-16      1.651309  -72.715728     -0.598550   
V3      284807.0 -1.379537e-15      1.516255  -48.325589     -0.890365   
V4      284807.0  2.074095e-15      1.415869   -5.683171     -0.848640   
V5      284807.0  9.604066e-16      1.380247 -113.743307     -0.691597   
V6      284807.0  1.487313e-15      1.332271  -26.160506     -0.768296   
V7      284807.0 -5.556467e-16      1.237094  -43.557242     -0.554076   
V8      284807.0  1.213481e-16      1.194353  -73.216718     -0.208630   
V9      284807.0 -2.406331e-15      1.098632  -13.434066     -0.643098   
V10     284807.0  2.239053e-15      1.088850  -24.588262     -0.535426   
V11     284807.0  1.673327e-15      1.020713   -4.797473     -0.762494   
V12     284807.0 -1.247012e-15      0.999201  -18.683715     -0.405571   
V13     284807.0  8.190001e-16      0.995274   -5.791881     -0.648539   
V14     284807.0  1.207294e-15      0.958596  -19.214325     -0.425574   
V15     284807.0  4.887456e-15      0.915316   -4.498945     -0.582884   
V16     284807.0  1.437716e-15      0.876253  -14.129855     -0.468037   
V17     284807.0 -3.772171e-16      0.849337  -25.162799     -0.483748   
V18     284807.0  9.564149e-16      0.838176   -9.498746     -0.498850   
V19     284807.0  1.039917e-15      0.814041   -7.213527     -0.456299   
V20     284807.0  6.406204e-16      0.770925  -54.497720     -0.211721   
V21     284807.0  1.654067e-16      0.734524  -34.830382     -0.228395   
V22     284807.0 -3.568593e-16      0.725702  -10.933144     -0.542350   
V23     284807.0  2.578648e-16      0.624460  -44.807735     -0.161846   
V24     284807.0  4.473266e-15      0.605647   -2.836627     -0.354586   
V25     284807.0  5.340915e-16      0.521278  -10.295397     -0.317145   
V26     284807.0  1.683437e-15      0.482227   -2.604551     -0.326984   
V27     284807.0 -3.660091e-16      0.403632  -22.565679     -0.070840   
V28     284807.0 -1.227390e-16      0.330083  -15.430084     -0.052960   
Amount  284807.0  8.834962e+01    250.120109    0.000000      5.600000   
Class   284807.0  1.727486e-03      0.041527    0.000000      0.000000   

                 50%            75%            max  
Time    84692.000000  139320.500000  172792.000000  
V1          0.018109       1.315642       2.454930  
V2          0.065486       0.803724      22.057729  
V3          0.179846       1.027196       9.382558  
V4         -0.019847       0.743341      16.875344  
V5         -0.054336       0.611926      34.801666  
V6         -0.274187       0.398565      73.301626  
V7          0.040103       0.570436     120.589494  
V8          0.022358       0.327346      20.007208  
V9         -0.051429       0.597139      15.594995  
V10        -0.092917       0.453923      23.745136  
V11        -0.032757       0.739593      12.018913  
V12         0.140033       0.618238       7.848392  
V13        -0.013568       0.662505       7.126883  
V14         0.050601       0.493150      10.526766  
V15         0.048072       0.648821       8.877742  
V16         0.066413       0.523296      17.315112  
V17        -0.065676       0.399675       9.253526  
V18        -0.003636       0.500807       5.041069  
V19         0.003735       0.458949       5.591971  
V20        -0.062481       0.133041      39.420904  
V21        -0.029450       0.186377      27.202839  
V22         0.006782       0.528554      10.503090  
V23        -0.011193       0.147642      22.528412  
V24         0.040976       0.439527       4.584549  
V25         0.016594       0.350716       7.519589  
V26        -0.052139       0.240952       3.517346  
V27         0.001342       0.091045      31.612198  
V28         0.011244       0.078280      33.847808  
Amount     22.000000      77.165000   25691.160000  

- **check the class inbalanced**

In [7]:
# check the class
data["Class"].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

- **2. Data Preparation**

In [8]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Separate features (X) and target (y)
X = data.iloc[:, :-1]
y = data["Class"]

# Normalize numerical features
scaler = StandardScaler()
X[["Amount", "Time"]] = scaler.fit_transform(X[["Amount", "Time"]])

# Split into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

In [9]:
X_train # training data ,input data 80%

Time        V1        V2        V3        V4        V5        V6  \
265518  1.413095  1.946747 -0.752526 -1.355130 -0.661630  1.502822  4.024933   
180305  0.624644  2.035149 -0.048880 -3.058693  0.247945  2.943487  3.298697   
42664  -1.129186 -0.991920  0.603193  0.711976 -0.992425 -0.825838  1.956261   
198723  0.796203  2.285718 -1.500239 -0.747565 -1.668119 -1.394143 -0.350339   
82325  -0.746606 -0.448747 -1.011440  0.115903 -3.454854  0.715771 -0.147490   
...          ...       ...       ...       ...       ...       ...       ...   
233802  1.113883  1.993864 -0.516866 -0.620118  0.129845 -0.285128  0.395044   
85418  -0.717019 -1.497933  0.657921  1.581568 -0.024286  0.584698  1.303031   
29062  -1.253217  1.069777  0.072105  0.496540  1.505318 -0.380277 -0.370243   
13766  -1.482496  1.280465  0.300586  0.333044  0.512720  0.065052 -0.145844   
17677  -1.389125 -0.598120  0.775041  1.823394  0.312991 -0.096171 -0.391452   

              V7        V8        V9  ...       V20       V21       V22  \
265518 -1.479661  1.139880  1.406819  ... -0.134435  0.076197  0.297537   
180305 -0.002192  0.674782  0.045826  ... -0.227279  0.038628  0.228197   
42664  -2.212603 -5.037523  0.000772  ...  1.280856 -2.798352  0.109526   
198723 -1.427984  0.010010 -1.118447  ... -0.490642 -0.139670  0.077013   
82325   0.504347 -0.113817 -0.044782  ... -0.275297 -0.243245 -0.173298   
...          ...       ...       ...  ...       ...       ...       ...   
233802 -0.822358  0.231591  0.995898  ... -0.174051  0.262526  0.884510   
85418   0.609212  0.135561  0.452745  ... -0.225079 -0.072452  0.299172   
29062   0.100551 -0.026687  0.319684  ... -0.149402 -0.061991 -0.044629   
13766  -0.145519 -0.091833  1.111187  ... -0.093783 -0.409934 -0.961351   
17677   0.499351  0.071224 -0.017496  ...  0.102259  0.060615  0.568083   

             V23       V24       V25       V26       V27       V28    Amount  
265518  0.307915  0.690980 -0.350316 -0.388907  0.077641 -0.032248 -0.323963  
180305  0.035542  0.707090  0.512885 -0.471198  0.002520 -0.069002 -0.341275  
42664  -0.436530 -0.932803  0.826684  0.913773  0.038049  0.185340  0.346835  
198723  0.208310 -0.538236 -0.278032 -0.162068  0.018045 -0.063005 -0.328841  
82325  -0.006692 -1.362383 -0.292234 -0.144622 -0.032580 -0.064194 -0.008994  
...          ...       ...       ...       ...       ...       ...       ...  
233802  0.099141  0.275689 -0.195404  0.623598 -0.032455 -0.058552 -0.329281  
85418   0.110048 -0.615980 -0.425883  0.263968 -0.448445  0.045178 -0.205340  
29062  -0.050485  0.400171  0.593314 -0.335160  0.031014  0.024886 -0.171636  
13766   0.033153 -0.560429  0.278428  0.089546 -0.059835 -0.005887 -0.349671  
17677  -0.084001  0.685003 -0.245859  0.356638  0.378580  0.206366 -0.325283  

[227845 rows x 30 columns]

In [10]:
X_test # testing data used to evaluate the model’s performance on unseen data. 20%

Time        V1        V2        V3        V4        V5        V6  \
263020  1.388689 -0.674466  1.408105 -1.110622 -1.328366  1.388996 -1.308439   
11378  -1.578646 -2.829816 -2.765149  2.537793 -1.074580  2.842559 -2.153536   
147283 -0.136621 -3.576495  2.318422  1.306985  3.263665  1.127818  2.865246   
219439  0.988041  2.060386 -0.015382 -1.082544  0.386019 -0.024331 -1.074935   
36939  -1.180778  1.209965  1.384303 -1.343531  1.763636  0.662351 -2.113384   
...          ...       ...       ...       ...       ...       ...       ...   
54199  -1.020990 -0.863057  0.225429  2.458855  0.613099  0.067149  1.716173   
184616  0.663243  1.397799 -1.426153 -0.369131  0.891825 -1.164153 -0.166657   
274532  1.500506 -1.047727  0.685141  0.195457 -3.583402  0.082922 -0.444060   
269819  1.452473  2.159972 -1.084234 -0.858819 -1.126188 -0.647032  0.234289   
64170  -0.921365  1.228059 -0.603264  0.246723 -0.414322 -1.170914 -0.895176   

              V7        V8        V9  ...       V20       V21       V22  \
263020  1.885879 -0.614233  0.311652  ...  0.394322  0.080084  0.810034   
11378  -1.795519 -0.250020  3.073504  ... -0.515765 -0.295555  0.109305   
147283  1.444125 -0.718922  1.874046  ...  2.034786 -1.060151  0.016867   
219439  0.207792 -0.338140  0.455091  ... -0.192024 -0.281684 -0.639426   
36939   0.854039 -0.475963 -0.629658  ...  0.009083 -0.164015 -0.328294   
...          ...       ...       ...  ...       ...       ...       ...   
54199  -0.254484  0.825754  0.407484  ... -0.207556  0.183183  0.695818   
184616 -0.320745 -0.023070  1.800818  ...  0.340394 -0.058383 -0.479606   
274532  0.261275  0.457403  1.225167  ... -0.222906 -0.120079 -0.388817   
269819 -1.164932  0.138244 -0.045273  ...  0.024393  0.327882  0.932738   
64170  -0.728271  0.027883 -0.480860  ...  0.050113  0.104995  0.060010   

             V23       V24       V25       V26       V27       V28    Amount  
263020 -0.224327  0.707899 -0.135837  0.045102  0.533837  0.291319 -0.261273  
11378  -0.813272  0.042996 -0.027660 -0.910247  0.110802 -0.511938 -0.305852  
147283 -0.132058 -1.483996 -0.296011  0.062823  0.552411  0.509764 -0.049095  
219439  0.331818 -0.067584 -0.283675  0.203529 -0.063621 -0.060077 -0.349271  
36939  -0.154631  0.619449  0.818998 -0.330525  0.046884  0.104527 -0.347232  
...          ...       ...       ...       ...       ...       ...       ...  
54199   0.101555 -0.683590 -0.505613 -0.343860  0.218972  0.127074 -0.209418  
184616  0.053303 -0.117313 -0.425722 -0.568239  0.001064  0.017812  0.836721  
274532 -0.095996 -1.059673 -0.392663 -0.781760  0.017347  0.145133 -0.287701  
269819  0.077597  0.119508 -0.090098 -0.101767 -0.002565 -0.056766 -0.273468  
64170   0.047525  0.265626  0.167280 -0.271841  0.024795  0.050174 -0.113744  

[56962 rows x 30 columns]

In [11]:
y_train # output for training data

265518    0
180305    0
42664     0
198723    0
82325     0
         ..
233802    0
85418     0
29062     0
13766     0
17677     0
Name: Class, Length: 227845, dtype: int64

In [12]:
y_test # output for testing data 

263020    0
11378     0
147283    0
219439    0
36939     0
         ..
54199     0
184616    0
274532    0
269819    0
64170     0
Name: Class, Length: 56962, dtype: int64

- **3 Handling Imbalanced**

In [13]:
from imblearn.over_sampling import SMOTE
smote = SMOTE(random_state=42)
X_train_balanced, y_train_balanced = smote.fit_resample(X_train, y_train)

- Everything is balanced now

In [14]:
print(y_train_balanced.value_counts())

Class
0    227451
1    227451
Name: count, dtype: int64


- **4. Training the Model (Random Forest)**

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, roc_auc_score

# Create and train the model
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train_balanced, y_train_balanced)

# Predict on test set
y_pred = model.predict(X_test)

- **5 Evaluate the model**

In [ ]:
# Evaluate
print(classification_report(y_test, y_pred))
print("ROC-AUC:", roc_auc_score(y_test, model.predict_proba(X_test)[:, 1]))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     56864
           1       0.85      0.84      0.84        98

    accuracy                           1.00     56962
   macro avg       0.92      0.92      0.92     56962
weighted avg       1.00      1.00      1.00     56962

ROC-AUC: 0.973103297664029


In [ ]:
from sklearn.metrics import balanced_accuracy_score
print("Balanced Accuracy:", balanced_accuracy_score(y_test, y_pred))

In [ ]:
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy of random forest is :", accuracy)

- **6 Conclusion**

###  Classification Report – Fraud Detection Model

### 🟩 Class 0 – **Non-Fraudulent Transactions**

- **Precision**: `1.00`  
  → 100% of transactions predicted as non-fraudulent are correct.

- **Recall**: `1.00`  
  → 100% of actual non-fraudulent transactions are correctly identified.

- **F1-score**: `1.00`  
  → Perfect balance of precision and recall.

- **Support**: `56,864`  
  → Number of non-fraudulent transactions in the test set.

**📝 Interpretation:**  
The model is excellent at identifying non-fraudulent transactions, which is expected given their overwhelming majority (**99.83%** of the data).

---

### 🟥 Class 1 – **Fraudulent Transactions**

- **Precision**: `0.85`  
  → 85% of transactions flagged as fraudulent are actually fraudulent.

- **Recall**: `0.84`  
  → 84% of actual fraudulent transactions are detected.

- **F1-score**: `0.84`  
  → Good balance between precision and recall.

- **Support**: `98`  
  → Number of fraudulent transactions in the test set (**0.17%**).

**📝 Interpretation:**  
The model performs well on the rare fraud class, detecting 84% of frauds with a reasonable false positive rate of 15%.

---

### ✅ Global Metrics

- **Accuracy**: `99.99`  
  ⚠️ Accuracy is misleading in imbalanced datasets: even if the model predicted every transaction as non-fraudulent, it would still achieve ~99.83% accuracy.
- **Balanced Accuracy**: `91%`  
  ⚠️ Accuracy is in imbalanced datasets is good
- **ROC-AUC**: `0.973`  
  Excellent ability of the model to **distinguish between fraudulent and non-fraudulent transactions**.

**📝 Interpretation:**  
The ROC curve shows that the model ranks transactions well, separating fraud from non-fraud effectively at various thresholds.

---
